forte/processors/base/data_augment_processor.py

```diff
@@ -46,68 +222,645 @@ class ReplacementDataAugmentProcessor(BaseDataAugmentProcessor):
     levels: character, word, sentence or document.
     """
     def __init__(self):
-        """
-        The replaced_spans records the entries replaced by new texts.
-        It consists of tuples (entry, new text) inserted by :func: replace.
-        The new text will be used for building new data pack.
-        """
         super().__init__()
-        self.replaced_spans: List[Tuple[Entry, str]] = []
+        # :attr:`_replaced_annos`: {datapack id: SortedList[span, new text]}
+        # It records the spans replaced by new texts.
+        # It is a map from datapack id to a list of tuples
+        # (span, new text) inserted by :func:`replace`.
+        # The new text will be used for building new data pack.
+        self._replaced_annos: DefaultDict[int, SortedList[Tuple[Span, str]]] = \
+            defaultdict(
+                lambda: SortedList([], key=lambda x: x[0])
+            )
+
+        # :attr:`_inserted_annos_pos_len`: {datapack id: Dict{position: length}}
+        # It records the position and length of inserted spans,
+        # mapping from datapack id to a dictionary (position -> length)
+        # inserted by :func:`insert`.
+        # The position is the index in the original datapack
+        # of insertion, and the length is the length of the inserted string.
+        self._inserted_annos_pos_len: DefaultDict[int, SortedDict[int, int]] = \
+            defaultdict(
+                lambda: SortedDict({})
+            )
 
-    def replace(self, replacement_op: TextReplacementOp, input: Annotation):
+        # :attr:`_deleted_annos_id`: {datapack id: Set[annotation tid]}
+        # It records the deleted span ids, mapping from datapack id
+        # to a set of annotation tids appended by :func:`delete`.
+        self._deleted_annos_id: DefaultDict[int, Set[int]] = defaultdict(set)
+
+        # :attr:`_data_pack_map`: {orig pack id: new pack id}
+        # It maintains a mapping from the pack id
+        # of the original pack to the pack id of augmented pack.
+        # It is used when copying the MultiPackLink and MultiPackGroup.
+        self._data_pack_map: Dict[int, int] = {}
+
+        # :attr:`_entry_maps`: {datapack id: Dict{orig tid, new tid}}
+        # It is a map for tracking the annotation ids
+        # before and after the auto align. It maps the
+        # original annotation tid to the new annotation tid.
+        # It is used when copying the Link/Group/MultiPackLink/MultiPackGroup.
+        self._entry_maps: Dict[int, Dict[int, int]] = {}
```

Comment:
I am not so sure about whether we can use `tid` safely here. We do have plans to make `tid` globally independent. But as for now, it is auto-incremented here: https://github.com/asyml/forte/blob/master/forte/data/container.py#L59

That means in a multipack, the `tid` itself cannot successfully refer to an Entry. In Forte, we have a class named Pointer to do that: https://github.com/asyml/forte/blob/3bb5e2cd96e948d44e3152414b80e677d407ccc9/forte/data/ontology/core.py#L445